query/queryable.d.ts

```diff
@@ -0,0 +1,146 @@
+/* Query Interfaces - Queryable */
+/* https://rdf.js.org/query-spec/ */
+
+import * as RDF from '../data-model';
+import { Bindings, Query, ResultStream } from './common';
+
+/**
+ * Context objects provide a way to pass additional bits information to the query engine when executing a query.
+ */
+export interface QueryContext<SourceType> {
+  /**
+   * An array of data sources the query engine must use.
+   */
+  sources?: [SourceType, ...SourceType[]];
+  /**
+   * The date that should be used by SPARQL operations such as NOW().
+   */
+  queryTimestamp?: Date;
+  /**
+   * Other options
+   */
+  [key: string]: any;
+}
+
+/**
+ * Context object in the case the passed query is a string.
+ */
+export interface QueryStringContext<SourceType> extends QueryContext<SourceType> {
+  /**
+   * The format in which the query string is defined.
+   * Defaults to { language: 'sparql', version: '1.1' }
+   */
+  queryFormat?: QueryFormat;
+  /**
+   * The baseIRI for parsing the query.
+   */
+  baseIRI?: string;
+}
+
+/**
+ * Context object in the case the passed query is an algebra object.
+ */
+export type QueryAlgebraContext<SourceType> = QueryContext<SourceType>;
+
+/**
+ * Represents a specific query format
+ */
+export interface QueryFormat {
+  /**
+   * The query language, e.g. 'sparql'.
+   */
+  language: string;
+  /**
+   * The version of the query language, e.g. '1.1'.
+   */
+  version: string;
+  /**
+   * An optional array of extensions on the query language.
+   * The representation of these extensions is undefined.
+   */
+  extensions?: string[];
+}
+
+/**
+ * Placeholder to represent SPARQL Algebra trees.
+ * Algebra typings are TBD. Reference implementations include:
+ * - https://www.npmjs.com/package/sparqlalgebrajs
+ */
+export type Algebra = any;
+
+/**
+ * Generic query engine interfaces.
+ * It allow engines to return any type of result object for any type of query.
```

Comment:
What is the use case for a non-SPARQL query engine?

Reply:
GraphQL(-LD) for instance.